### DISCUSSION TITLE
Sharing SPIRAM with Micropython (GC question)

### DISCUSSION THREAD

**Discussion by BrunoESP32** [micropython/micropython#13507] [ESP32] [answered]
Hi,

I am working on a C application and require more RAM than what is available internally on the ESP32-S3. My intention is to utilize a portion of the SPIRAM with Micropython.

In the main.c file, I came across the following code:

```
#elif CONFIG_ESP32S2_SPIRAM_SUPPORT || CONFIG_ESP32S3_SPIRAM_SUPPORT
    size_t esp_spiram_size = esp_spiram_get_size();
    if (esp_spiram_size > 0) {
        mp_task_heap = (void *)SOC_EXTRAM_DATA_HIGH - esp_spiram_size;
        mp_task_heap_size = esp_spiram_size;
    }
#endif
```

If I reduce Micropython's allocation and use the remaining memory, will GC catch it? The following line made me doubt:

```
gc_init(mp_task_heap, mp_task_heap + mp_task_heap_size);
```
Is it safe to use memory beyond ```mp_task_heap + mp_task_heap_size```?

Thanks.

**Comment by glenn20** [comment, marked as answer]
If you reduce `mp_task_heap_size` you can reserve the memory above for your C application/module's use. GC wil ignore any memory outside the start and end values passed to gc_init().

An alternative approach is to allocate a block of memory from the micropython heap and hold a reference to it as long as you need it so that it does not get gc-ed. If it is a large block, you should allocate soon after boot to avoid heap fragmentation. This way you can dynamically set the size of memory for your module (if that is useful for your use case).

Another approach is to allocate the memory in python code as a bytearray (or other type of array) and pass that to your C module to use. This has the convenience that you can easily inspect the memory from python code while developing and debugging (although there are other ways to expose that memory to python.

**Reply by BrunoESP32** [reply]
Thanks for the replay and the tips, [glenn20](https://github.com/glenn20)!

Yes! I will allocate before initialization to avoid heap fragmentation. Thank you =)

**Comment by jimmo** [comment]
@BrunoESP32 All of this has changed recently. I think you're looking at an older version of the codebase.

See https://github.com/orgs/micropython/discussions/12316

**Reply by BrunoESP32** [reply]
Thanks for the replay, [jimmo](https://github.com/jimmo)!

Yes, I'm using an older version (1.20.0), as I was initially developing with it, I haven't looked at the latest versions, but I will! Thank you very much!

**Reply by BrunoESP32** [reply]
Hi, [jimmo](https://github.com/jimmo)!

I'm upgrading the firmware version I was using, and another question arose about the GC and this new approach to using SPIRAM memory. Can I use heap_caps_malloc on newer versions of Micropython firmware? Will the GC ignore allocations that were not done by micropython?

**Reply by jimmo** [reply]
@BrunoESP32 Yes you can use any IDF malloc functions.

This is basically the idea of these changes, by not pre-allocating up-front the entire MicroPython heap, we allow a lot more flexibility for C code to malloc/free as required.

**Reply by BrunoESP32** [reply]
Thanks, [jimmo](https://github.com/jimmo)!

**Comment by picpic020960** [comment]
@glenn20 write 'This has the convenience that you can easily inspect the memory from python code while developing and debugging (although there are other ways to expose that memory to python.'
Examples ?
